resources/session04/mashup_6.py

```diff
@@ -0,0 +1,263 @@
+    return resp.text
+
+
+def parse_source(html):
+    parsed = BeautifulSoup(html, 'html5lib')
+    return parsed
+
+
+def load_inspection_page(name):
+    file_path = pathlib.Path(name)
+    return file_path.read_text(encoding='utf8')
+
+
+def restaurant_data_generator(html):
+    id_finder = re.compile(r'PR[\d]+~')
+    return html.find_all('div', id=id_finder)
+
+
+def has_two_tds(elem):
+    is_tr = elem.name == 'tr'
+    td_children = elem.find_all('td', recursive=False)
+    has_two = len(td_children) == 2
+    return is_tr and has_two
+
+
+def clean_data(td):
+    return td.text.strip(" \n:-")
+
+
+def extract_restaurant_metadata(elem):
+    restaurant_data_rows = elem.find('tbody').find_all(
+        has_two_tds, recursive=False
+    )
+    rdata = {}
+    current_label = ''
+    for data_row in restaurant_data_rows:
+        key_cell, val_cell = data_row.find_all('td', recursive=False)
+        new_label = clean_data(key_cell)
+        current_label = new_label if new_label else current_label
+        rdata.setdefault(current_label, []).append(clean_data(val_cell))
+    return rdata
+
+
+def is_inspection_data_row(elem):
+    is_tr = elem.name == 'tr'
+    if not is_tr:
+        return False
+    td_children = elem.find_all('td', recursive=False)
+    has_four = len(td_children) == 4
+    this_text = clean_data(td_children[0]).lower()
+    contains_word = 'inspection' in this_text
+    does_not_start = not this_text.startswith('inspection')
+    return is_tr and has_four and contains_word and does_not_start
+
+
+def get_score_data(elem):
+    inspection_rows = elem.find_all(is_inspection_data_row)
+    samples = len(inspection_rows)
+    total = 0
+    high_score = 0
+    average = 0
+    for row in inspection_rows:
+        strval = clean_data(row.find_all('td')[2])
+        try:
+            intval = int(strval)
+        except (ValueError, TypeError):
+            samples -= 1
+        else:
+            total += intval
+            high_score = intval if intval > high_score else high_score
+
+    if samples:
+        average = total/float(samples)
+    data = [
+        (u'Average Score', average),
+        (u'High Score', high_score),
+        (u'Total Inspections', samples)
+    ]
+    return data
+
+
+def result_generator(count):
+    use_params = {
+        'Inspection_Start': '01/26/2015',
+        'Inspection_End': '201/26/2016',
+        'Zip_Code': '98101'
+    }
+    # html = get_inspection_page(**use_params)
+    sorting_key, marker_type = check_sorting()
+    html = load_inspection_page('inspection_page.html')
+    parsed = parse_source(html)
+    content_col = parsed.find("td", id="contentcol")
+    data_list = restaurant_data_generator(content_col)
+    for data_div in data_list[:count]:
+        metadata = extract_restaurant_metadata(data_div)
+        inspection_data = get_score_data(data_div)
+        metadata.update(inspection_data)
+        # sorting_value = str(metadata.get(sorting_key))
+        yield metadata
```

Comment:
This function is going to hand us back one converted restaurant record at a time, and it will only hand back `count` number of records.  We need to change the order of our restaurants in here, or there will be no way for us to sort them later.  

The idea here, in psuedocode, would be this:

```python
data_list = restaurant_data_generator(...)
all_metadata = []
for data_div in data_list:
    # build metadata as before
    all_metadata.append(metadata)
for record in sorted(metadata, key=sorting_function, reverse=true_or_false)[:count]
    yield record
```

This means that you'd need to update this function to take some value for sort order and for sorting value as arguments.  Then you could use them in the function as above.